Gems/ROS2/Code/Source/Lidar/Publishing/PointCloudMessageFormat.h

```diff
@@ -0,0 +1,166 @@
+#pragma once
+
+#include "AzFramework/Physics/Common/PhysicsSceneQueries.h"
+
+#include <AzCore/Outcome/Outcome.h>
+#include <AzCore/base.h>
+#include <AzCore/std/containers/unordered_map.h>
+#include <AzCore/std/smart_ptr/unique_ptr.h>
+#include <AzCore/std/string/regex.h>
+#include <ROS2/Lidar/RaycastResults.h>
+#include <rclcpp/publisher.hpp>
+#include <sensor_msgs/msg/point_cloud2.hpp>
+#include <vision_msgs/msg/label_info.hpp>
+
+namespace ROS2
+{
+    enum class FieldFlags : uint16_t
+    {
+        // clang-format off
+        None                    = 0,
+        PositionXYZF32          = 1,
+        IntensityF32            = 1 << 1,
+        TU32                    = 1 << 2, // ?
+        ReflectivityU16         = 1 << 3,
+        RingU16                 = 1 << 4,
+        AmbientU16              = 1 << 5,
+        RangeU32                = 1 << 6,
+        SegmentationData96      = 1 << 7,
+        Padding8                = 1 << 8,
+        Padding16               = 1 << 9,
+        Padding32               = 1 << 10,
+        All                     = 0b0000'0111'1111'1111,
+        // clang-format on
+    };
+
+    bool IsPadding(FieldFlags fieldFlag);
+    size_t GetFieldByteSize(FieldFlags fieldFlag);
+    size_t GetActualFieldCount(FieldFlags fieldFlag);
+    RaycastResultFlags GetFieldProvider(FieldFlags fieldFlag);
+    AZStd::optional<AZStd::string> GetDefaultFieldName(FieldFlags fieldFlag);
+    AZStd::optional<AZStd::string> GetDatatypeString(FieldFlags fieldFlag);
+
+    template<FieldFlags fieldFlag>
+    struct FieldTraits;
+
+    template<>
+    struct FieldTraits<FieldFlags::PositionXYZF32>
+    {
+        struct Position
+        {
+            float m_x, m_y, m_z;
+        };
+
+        using Type = Position;
+        static constexpr Type DefaultValue = { .m_x = 0.0f, .m_y = 0.0f, .m_z = 0.0f };
+        static constexpr RaycastResultFlags Provider = RaycastResultFlags::Point;
+    };
+
+    template<>
+    struct FieldTraits<FieldFlags::IntensityF32>
+    {
+        using Type = float;
+        static constexpr Type DefaultValue = 0.0f;
+        static constexpr RaycastResultFlags Provider = RaycastResultFlags::Intensity;
+    };
+
+    template<>
+    struct FieldTraits<FieldFlags::TU32>
+    {
+        using Type = AZ::u32;
+        static constexpr Type DefaultValue = 0U;
+        static constexpr RaycastResultFlags Provider = RaycastResultFlags::None;
+    };
+
+    template<>
+    struct FieldTraits<FieldFlags::ReflectivityU16>
+    {
+        using Type = float;
+        static constexpr Type DefaultValue = 0.0f;
+        static constexpr RaycastResultFlags Provider = RaycastResultFlags::None;
+    };
+
+    template<>
+    struct FieldTraits<FieldFlags::RingU16>
+    {
+        using Type = AZ::u16;
+        static constexpr Type DefaultValue = 0U;
+        static constexpr RaycastResultFlags Provider = RaycastResultFlags::None;
```

Comment:
Ring (channel) can be returned from RGL, but requires additional configuration through API. We can discuss this if it will be necessary.

Reply:
I would consider adding this as a raycast result after the rest of the PR is ready (possibly as a separate PR).